public/Aframe/samples/loading-scene-1/components.js

```diff
@@ -0,0 +1,127 @@
+AFRAME.registerComponent("collision-detect", {
+    init: function () {
+        this.el.addEventListener("collide", function (e) {
+            console.log("Player has collided with ", e.detail.body.el);
+            e.detail.target.el; // Original entity (playerEl).
+            e.detail.body.el; // Other entity, which playerEl touched.
+            e.detail.contact; // Stats about the collision (CANNON.ContactEquation).
+            e.detail.contact.ni; // Normal (direction) of the collision (CANNON.Vec3).
+        });
+    },
+});
+
+AFRAME.registerComponent("detect-button", {
+    init: function () {
+        var data = this.data;
+        var el = this.el;
+
+        el.addEventListener("mouseenter", function () {
+            el.setAttribute(
+                "animation__button-enter",
+                "property: scale; to: 1.1 1.1 1.1; dur: 500; dir:alternate; startEvents: mouseenter"
+            );
+        });
+
+        el.addEventListener("mouseleave", function () {
+            el.setAttribute(
+                "animation__button-leave",
+                "property: scale; to: 1 1 1; startEvents: mouseleave"
+            );
+        });
+
+        el.addEventListener("mousedown", function () {
+            el.setAttribute(
+                "animation__button-pressed",
+                "property: object3D.position.y; to: -5.5; dir: alternate; dur: 50; startEvents: mousedown"
+            );
+
+            el.emit("buttondown", false);
+        });
+
+        el.addEventListener("mouseup", function () {
+            el.setAttribute(
+                "animation__button-released",
+                "property: object3D.position.y; to: -5; dir: alternate; dur: 100; startEvents: mouseup"
+            );
+        });
+    },
+});
+
+function addBall(x, y, z) {
+    let container = document.querySelector("#container");
+    let colors = [
+        "red",
+        "orange",
+        "yellow",
+        "green",
+        "blue",
+        "purple",
+        "hotpink",
+    ];
+
+    container.innerHTML += `<a-sphere collision-detect id="ball" click-drag dynamic-body position="${x} ${y} ${z}" radius="0.5" color="${
```

Comment:
While this way works, I think it would be better to use javascript to build the sphere element and append to the container instead.

So instead of `container.innerHTML = ... `, you would have something like we do in this `setupEnv() method` in [this example](https://github.com/uwblueprint/cdc-backend/blob/c0ef19b59a86e6ba2788ba04673fdda609af6bba/samples/popup/popup.html#L55).

Reply:
Funny you mention that. I actually tried that form, good to know that's the "better" way of doing things. If I wanted to add a custom component to an entity in this way, would the syntax be `entity.setAttribute("custom-name")`? I think that was the only reason I didn't switch to using that form.

Reply:
Yep, that is the way. Here is some aframe documentation on uses for setAttribute that might help https://aframe.io/docs/1.2.0/core/entity.html#setattribute-componentname-value-propertyvalue-clobber